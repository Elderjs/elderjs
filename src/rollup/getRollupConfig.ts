import svelte from 'rollup-plugin-svelte';
import { nodeResolve } from '@rollup/plugin-node-resolve';
import commonjs from '@rollup/plugin-commonjs';
import { terser } from 'rollup-plugin-terser';
import babel from 'rollup-plugin-babel';
import multiInput from 'rollup-plugin-multi-input';
import replace from '@rollup/plugin-replace';
import json from '@rollup/plugin-json';
import glob from 'glob';
import path from 'path';
import fs from 'fs-extra';
import del from 'del';
import defaultsDeep from 'lodash.defaultsdeep';
import { getElderConfig } from '../index';
import { getDefaultRollup } from '../utils/validations';
import getPluginLocations from '../utils/getPluginLocations';
import elderSvelte from './rollupPlugin';

const production = process.env.NODE_ENV === 'production' || !process.env.ROLLUP_WATCH;
const elderJsDir = path.resolve(process.cwd(), './node_modules/@elderjs/elderjs/');

const babelIE11 = babel({
  extensions: ['.js', '.mjs', '.html', '.svelte'],
  runtimeHelpers: true,
  exclude: ['node_modules/@babel/**', 'node_modules/core-js/**', /\/core-js\//],
  presets: [
    [
      '@babel/preset-env',
      {
        targets: {
          browsers: ['> 0.25%', 'not dead', 'IE 11'],
        },
        useBuiltIns: 'usage',
        forceAllTransforms: true,
        corejs: {
          version: 3.6,
          proposals: true,
        },
      },
    ],
  ],
});

export function createBrowserConfig({
  input,
  output,
  multiInputConfig,
  svelteConfig,
  replacements = {},
  ie11 = false as boolean,
  distElder,
  distDir,
  rootDir,
}) {
  const toReplace = {
    'process.env.componentType': "'browser'",
    'process.env.NODE_ENV': JSON.stringify(process.env.NODE_ENV),
    ...replacements,
  };

  const config = {
    cache: true,
    treeshake: production,
    input,
    output,
    plugins: [
      replace(toReplace),
      json(),
      elderSvelte({ svelteConfig, type: 'client', distElder, distDir, rootDir, legacy: ie11 }),
      nodeResolve({
        browser: true,
        dedupe: ['svelte', 'core-js'],
        preferBuiltins: true,
        rootDir: ie11 ? elderJsDir : process.cwd(),
      }),
      commonjs({ sourceMap: !production }),
    ],
  };

  // bundle splitting.
  if (multiInputConfig) {
    config.plugins.unshift(multiInputConfig);
  }

  // ie11 babel
  if (ie11) {
    config.plugins.push(babelIE11);
  }

  // if is production let's babelify everything and minify it.
  if (production) {
    // don't babel if it has been done
    if (!ie11) {
      config.plugins.push(
        babel({
          extensions: ['.js', '.mjs', '.cjs', '.html', '.svelte'],
          include: ['node_modules/**', 'src/**'],
          exclude: ['node_modules/@babel/**'],
          runtimeHelpers: true,
        }),
      );
    }

    // terser on prod
    config.plugins.push(terser());
  }

  return config;
}

export function createSSRConfig({
  input,
  output,
  svelteConfig,
  replacements = {},
  multiInputConfig,
  rootDir,
  distElder,
  distDir,
}) {
  const toReplace = {
    'process.env.componentType': "'server'",
    'process.env.NODE_ENV': JSON.stringify(process.env.NODE_ENV),
    ...replacements,
  };

  const config = {
    cache: true,
    treeshake: production,
    input,
    output,
    plugins: [
      replace(toReplace),
      json(),
      elderSvelte({ svelteConfig, type: 'ssr', distElder, distDir, rootDir }),
      nodeResolve({
        browser: false,
        dedupe: ['svelte'],
      }),
      commonjs({ sourceMap: true }),
      production && terser(),
    ],
  };
  // if we are bundle splitting include them.
  if (multiInputConfig) {
    config.plugins.unshift(multiInputConfig);
  }

  return config;
}

export default function getRollupConfig(options) {
  const defaultOptions = getDefaultRollup();
  const { svelteConfig, replacements, dev } = defaultsDeep(options, defaultOptions);
  const elderConfig = getElderConfig();
  const { $$internal, distDir, srcDir, rootDir, legacy } = elderConfig;
  const { ssrComponents, clientComponents, distElder } = $$internal;
  const relSrcDir = srcDir.replace(rootDir, '').substr(1);

  console.log(`Elder.js using rollup in ${production ? 'production' : 'development'} mode.`);

  const configs = [];

  // clear out components so there are no conflicts due to hashing.
  del.sync([`${ssrComponents}*`, `${clientComponents}*`]);
  // Add ElderJs Peer deps to public if they exist.
  [['./node_modules/intersection-observer/intersection-observer.js', './static/intersection-observer.js']].forEach(
    (dep) => {
      if (!fs.existsSync(path.resolve(rootDir, dep[0]))) {
        throw new Error(`Elder.js peer dependency not found at ${dep[0]}`);
      }
      configs.push({
        input: dep[0],
        output: [
          {
            file: path.resolve(distDir, dep[1]),
            format: 'iife',
            name: dep[1],
            plugins: [terser()],
          },
        ],
      });
    },
  );

  const { paths: pluginPaths, files: pluginFiles } = getPluginLocations(elderConfig);
  const pluginGlobs = pluginPaths.map((plugin) => `${plugin}*.svelte`);

  const components = fs.existsSync(path.resolve(srcDir, `./components/`))
    ? [
        ...new Set([
          ...glob.sync(path.resolve(srcDir, './components/*/*.svelte')),
          ...glob.sync(path.resolve(srcDir, './components/*.svelte')),
        ]),
      ]
    : [];

  configs.push(
    createSSRConfig({
      input: [
        `${relSrcDir}/layouts/*.svelte`,
        `${relSrcDir}/routes/**/*.svelte`,
        `${relSrcDir}/components/**/*.svelte`,
        ...pluginGlobs,
      ],
      output: {
        dir: ssrComponents,
        format: 'cjs',
        exports: 'auto',
        sourcemap: !production ? 'inline' : false,
      },
      multiInputConfig: multiInput({
        relative: 'src/',
        // transformOutputPath: (output) => ssrOutputPath(output),
      }),
      svelteConfig,
      replacements,
      rootDir,
      distElder,
      distDir,
    }),
  );

  configs.push(
    createBrowserConfig({
      input: [`${relSrcDir}/components/**/*.svelte`],
      output: [
        {
          dir: clientComponents,
          sourcemap: !production ? 'inline' : false,
          format: 'esm',
          entryFileNames: '[name].[hash].js',
        },
      ],
      multiInputConfig: multiInput({
        relative: 'src/',
        // transformOutputPath: (output) => `${path.basename(output)}`,
      }),
      svelteConfig,
      replacements,
      distElder,
      distDir,
      rootDir,
    }),
  );

  // legacy is only done on production or not split modes.
  if (legacy && production) {
    [...components, ...pluginFiles].forEach((cv) => {
      const file = cv.replace(`${rootDir}/`, '');
      const parsed = path.parse(cv);
      configs.push(
        createBrowserConfig({
          input: file,
          output: [
            {
              name: `___elderjs_${parsed.name}`,
              entryFileNames: '[name].[hash].js',
              dir: `${clientComponents}${path.sep}iife${path.sep}`,
              sourcemap: !production,
              format: 'iife',
            },
          ],
          svelteConfig,
          replacements,
          multiInputConfig: false,
          ie11: true,
          distElder,
          distDir,
          rootDir,
        }),
      );
    });
  }

  return configs;
}
