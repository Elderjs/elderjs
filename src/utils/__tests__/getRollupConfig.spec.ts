/* eslint-disable global-require */
import multiInput from 'rollup-plugin-multi-input';
import path from 'path';
import { createBrowserConfig, createSSRConfig } from '../getRollupConfig';

describe('#getRollupConfig', () => {
  beforeEach(() => {
    jest.resetModules();
  });

  it('createBrowserConfig works', () => {
    [true, false].forEach((sourcemap) => {
      expect(
        createBrowserConfig({
          input: [`./components/*/*.svelte`],
          output: {
            dir: './public/dist/svelte/',
            entryFileNames: 'entry[name]-[hash].js',
            sourcemap,
            format: 'system',
          },
          multiInputConfig: multiInput({
            // TODO: test with false
            relative: `./components`,
            transformOutputPath: (output) => `${path.basename(output)}`,
          }),
          svelteConfig: {},
        }),
      ).toEqual({
        cache: true,
        input: ['./components/*/*.svelte'],
        output: {
          dir: './public/dist/svelte/',
          entryFileNames: 'entry[name]-[hash].js',
          format: 'system',
          sourcemap,
        },
        plugins: [
          {
            options: expect.any(Function),
            pluginName: 'rollup-plugin-multi-input',
          },
          {
            name: 'replace',
            renderChunk: expect.any(Function),
            transform: expect.any(Function),
          },
          {
            name: 'json',
            transform: expect.any(Function),
          },
          {
            generateBundle: expect.any(Function),
            load: expect.any(Function),
            name: 'svelte',
            resolveId: expect.any(Function),
            transform: expect.any(Function),
          },
          {
            name: 'rollup-plugin-external-globals',
            transform: expect.any(Function),
          },
          {
            buildStart: expect.any(Function),
            generateBundle: expect.any(Function),
            getPackageInfoForId: expect.any(Function),
            load: expect.any(Function),
            name: 'node-resolve',
            resolveId: expect.any(Function),
          },
          {
            buildStart: expect.any(Function),
            load: expect.any(Function),
            name: 'commonjs',
            resolveId: expect.any(Function),
            transform: expect.any(Function),
          },
          {
            load: expect.any(Function),
            name: 'babel',
            resolveId: expect.any(Function),
            transform: expect.any(Function),
          },
          {
            name: 'terser',
            renderChunk: expect.any(Function),
          },
        ],
        treeshake: true,
      });
    });
  });

  it('createBrowserConfig multiInputConfig = false', () => {
    expect(
      createBrowserConfig({
        input: [`./components/*/*.svelte`],
        output: {
          dir: './public/dist/svelte/',
          entryFileNames: 'entry[name]-[hash].js',
          sourcemap: true,
          format: 'system',
        },
        svelteConfig: {},
      }).plugins,
    ).toEqual([
      {
        name: 'replace',
        renderChunk: expect.any(Function),
        transform: expect.any(Function),
      },
      {
        name: 'json',
        transform: expect.any(Function),
      },
      {
        generateBundle: expect.any(Function),
        load: expect.any(Function),
        name: 'svelte',
        resolveId: expect.any(Function),
        transform: expect.any(Function),
      },
      {
        name: 'rollup-plugin-external-globals',
        transform: expect.any(Function),
      },
      {
        buildStart: expect.any(Function),
        generateBundle: expect.any(Function),
        getPackageInfoForId: expect.any(Function),
        load: expect.any(Function),
        name: 'node-resolve',
        resolveId: expect.any(Function),
      },
      {
        buildStart: expect.any(Function),
        load: expect.any(Function),
        name: 'commonjs',
        resolveId: expect.any(Function),
        transform: expect.any(Function),
      },
      {
        load: expect.any(Function),
        name: 'babel',
        resolveId: expect.any(Function),
        transform: expect.any(Function),
      },
      {
        name: 'terser',
        renderChunk: expect.any(Function),
      },
    ]);
  });

  it('createSSRConfig works', () => {
    expect(
      createSSRConfig({
        input: [`./components/*/*.svelte`],
        output: {
          dir: './___ELDER___/compiled/',
          format: 'cjs',
          exports: 'auto',
        },
        multiInputConfig: multiInput({
          relative: `./components`,
          transformOutputPath: (output) => `${path.basename(output)}`,
        }),
        svelteConfig: {
          preprocess: [
            {
              style: ({ content }) => {
                return content.toUpperCase();
              },
            },
          ],
        },
      }),
    ).toEqual({
      cache: true,
      input: ['./components/*/*.svelte'],
      output: {
        dir: './___ELDER___/compiled/',
        exports: 'auto',
        format: 'cjs',
      },
      plugins: [
        {
          options: expect.any(Function),
          pluginName: 'rollup-plugin-multi-input',
        },
        {
          name: 'replace',
          renderChunk: expect.any(Function),
          transform: expect.any(Function),
        },
        {
          name: 'json',
          transform: expect.any(Function),
        },
        {
          generateBundle: expect.any(Function),
          load: expect.any(Function),
          name: 'svelte',
          resolveId: expect.any(Function),
          transform: expect.any(Function),
        },
        {
          buildStart: expect.any(Function),
          generateBundle: expect.any(Function),
          getPackageInfoForId: expect.any(Function),
          load: expect.any(Function),
          name: 'node-resolve',
          resolveId: expect.any(Function),
        },
        {
          buildStart: expect.any(Function),
          load: expect.any(Function),
          name: 'commonjs',
          resolveId: expect.any(Function),
          transform: expect.any(Function),
        },
        {
          generateBundle: expect.any(Function),
          name: 'css',
          transform: expect.any(Function),
        },
        {
          name: 'terser',
          renderChunk: expect.any(Function),
        },
      ],
      treeshake: true,
    });
  });

  it('createSSRConfig multiInputConfig = false', () => {
    expect(
      createSSRConfig({
        input: [`./components/*/*.svelte`],
        output: {
          dir: './___ELDER___/compiled/',
          format: 'cjs',
          exports: 'auto',
        },
        svelteConfig: {
          preprocess: [
            {
              style: ({ content }) => {
                return content.toUpperCase();
              },
            },
          ],
        },
      }).plugins,
    ).toEqual([
      {
        name: 'replace',
        renderChunk: expect.any(Function),
        transform: expect.any(Function),
      },
      {
        name: 'json',
        transform: expect.any(Function),
      },
      {
        generateBundle: expect.any(Function),
        load: expect.any(Function),
        name: 'svelte',
        resolveId: expect.any(Function),
        transform: expect.any(Function),
      },
      {
        buildStart: expect.any(Function),
        generateBundle: expect.any(Function),
        getPackageInfoForId: expect.any(Function),
        load: expect.any(Function),
        name: 'node-resolve',
        resolveId: expect.any(Function),
      },
      {
        buildStart: expect.any(Function),
        load: expect.any(Function),
        name: 'commonjs',
        resolveId: expect.any(Function),
        transform: expect.any(Function),
      },
      {
        generateBundle: expect.any(Function),
        name: 'css',
        transform: expect.any(Function),
      },
      {
        name: 'terser',
        renderChunk: expect.any(Function),
      },
    ]);
  });

  it('getPluginPaths works', () => {
    jest.mock('glob', () => ({
      sync: jest
        .fn()
        .mockImplementationOnce(() => ['a-file1.svelte', 'a-file2.svelte'])
        .mockImplementationOnce(() => ['b-file1.svelte', 'b-file2.svelte']),
    }));

    jest.mock('fs-extra', () => ({
      existsSync: jest
        .fn()
        .mockImplementationOnce(() => true) // first plugin from src
        .mockImplementationOnce(() => false) // 2nd from node modules
        .mockImplementationOnce(() => true),
    }));

    expect(
      // @ts-ignore
      require('../getRollupConfig').getPluginPaths({
        srcDir: './src',
        rootDir: './',
        plugins: {
          pluginA: {},
          pluginB: {},
        },
      }),
    ).toEqual([process.cwd() + '/src/plugins/pluginA/', process.cwd() + '/node_modules/pluginB/']);
  });

  it('getRollupConfig as a whole works', () => {
    jest.mock('../getConfig', () => () => ({
      $$internal: {
        clientComponents: 'test/public/svelte',
        ssrComponents: 'test/___ELDER___/compiled',
      },
      distDir: './dist',
      srcDir: './src',
      rootDir: './',
      plugins: {
        pluginA: {},
        pluginB: {},
      },
    }));

    jest.mock('del');
    jest.mock('fs-extra', () => ({
      existsSync: jest.fn().mockImplementation(() => true),
    }));
    jest.mock('glob', () => ({
      sync: jest
        .fn()
        .mockImplementationOnce(() => ['route1.svelte', 'route2.svelte'])
        .mockImplementationOnce(() => ['layout1.svelte', 'layout2.svelte'])
        .mockImplementationOnce(() => ['pluginA.svelte', 'pluginB.svelte'])
        .mockImplementationOnce(() => ['foo.svelte', 'bar.svelte']),
    }));

    const svelteConfig = {
      preprocess: [
        {
          style: ({ content }) => {
            return content.toUpperCase();
          },
        },
      ],
    };
    expect(require('../getRollupConfig').default({ svelteConfig })).toStrictEqual(
      expect.arrayContaining([
        expect.objectContaining({
          input: './node_modules/intersection-observer/intersection-observer.js',
        }),
      ]),
    );
  });
});
