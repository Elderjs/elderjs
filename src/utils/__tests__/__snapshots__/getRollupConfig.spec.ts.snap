// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#getRollupConfig getRollupConfig as a whole works 1`] = `
Array [
  Object {
    "input": "./node_modules/intersection-observer/intersection-observer.js",
    "output": Array [
      Object {
        "file": "/Users/lukewatts/oss/elderjs/dist/static/intersection-observer.js",
        "format": "iife",
        "name": "./static/intersection-observer.js",
        "plugins": Array [
          Object {
            "name": "terser",
            "renderChunk": [Function],
          },
        ],
      },
    ],
  },
  Object {
    "input": "./node_modules/systemjs/dist/s.min.js",
    "output": Array [
      Object {
        "file": "/Users/lukewatts/oss/elderjs/dist/static/s.min.js",
        "format": "iife",
        "name": "./static/s.min.js",
        "plugins": Array [
          Object {
            "name": "terser",
            "renderChunk": [Function],
          },
        ],
      },
    ],
  },
  Object {
    "cache": true,
    "input": "route1.svelte",
    "output": Object {
      "dir": "test/___ELDER___/compiled",
      "exports": "auto",
      "format": "cjs",
    },
    "plugins": Array [
      Object {
        "name": "replace",
        "renderChunk": [Function],
        "transform": [Function],
      },
      Object {
        "name": "json",
        "transform": [Function],
      },
      Object {
        "generateBundle": [Function],
        "load": [Function],
        "name": "svelte",
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "buildStart": [Function],
        "generateBundle": [Function],
        "getPackageInfoForId": [Function],
        "load": [Function],
        "name": "node-resolve",
        "resolveId": [Function],
      },
      Object {
        "buildStart": [Function],
        "load": [Function],
        "name": "commonjs",
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "generateBundle": [Function],
        "name": "css",
        "transform": [Function],
      },
      Object {
        "name": "terser",
        "renderChunk": [Function],
      },
    ],
    "treeshake": true,
  },
  Object {
    "cache": true,
    "input": "route2.svelte",
    "output": Object {
      "dir": "test/___ELDER___/compiled",
      "exports": "auto",
      "format": "cjs",
    },
    "plugins": Array [
      Object {
        "name": "replace",
        "renderChunk": [Function],
        "transform": [Function],
      },
      Object {
        "name": "json",
        "transform": [Function],
      },
      Object {
        "generateBundle": [Function],
        "load": [Function],
        "name": "svelte",
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "buildStart": [Function],
        "generateBundle": [Function],
        "getPackageInfoForId": [Function],
        "load": [Function],
        "name": "node-resolve",
        "resolveId": [Function],
      },
      Object {
        "buildStart": [Function],
        "load": [Function],
        "name": "commonjs",
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "generateBundle": [Function],
        "name": "css",
        "transform": [Function],
      },
      Object {
        "name": "terser",
        "renderChunk": [Function],
      },
    ],
    "treeshake": true,
  },
  Object {
    "cache": true,
    "input": "layout1.svelte",
    "output": Object {
      "dir": "test/___ELDER___/compiled",
      "exports": "auto",
      "format": "cjs",
    },
    "plugins": Array [
      Object {
        "name": "replace",
        "renderChunk": [Function],
        "transform": [Function],
      },
      Object {
        "name": "json",
        "transform": [Function],
      },
      Object {
        "generateBundle": [Function],
        "load": [Function],
        "name": "svelte",
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "buildStart": [Function],
        "generateBundle": [Function],
        "getPackageInfoForId": [Function],
        "load": [Function],
        "name": "node-resolve",
        "resolveId": [Function],
      },
      Object {
        "buildStart": [Function],
        "load": [Function],
        "name": "commonjs",
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "generateBundle": [Function],
        "name": "css",
        "transform": [Function],
      },
      Object {
        "name": "terser",
        "renderChunk": [Function],
      },
    ],
    "treeshake": true,
  },
  Object {
    "cache": true,
    "input": "layout2.svelte",
    "output": Object {
      "dir": "test/___ELDER___/compiled",
      "exports": "auto",
      "format": "cjs",
    },
    "plugins": Array [
      Object {
        "name": "replace",
        "renderChunk": [Function],
        "transform": [Function],
      },
      Object {
        "name": "json",
        "transform": [Function],
      },
      Object {
        "generateBundle": [Function],
        "load": [Function],
        "name": "svelte",
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "buildStart": [Function],
        "generateBundle": [Function],
        "getPackageInfoForId": [Function],
        "load": [Function],
        "name": "node-resolve",
        "resolveId": [Function],
      },
      Object {
        "buildStart": [Function],
        "load": [Function],
        "name": "commonjs",
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "generateBundle": [Function],
        "name": "css",
        "transform": [Function],
      },
      Object {
        "name": "terser",
        "renderChunk": [Function],
      },
    ],
    "treeshake": true,
  },
  Object {
    "cache": true,
    "input": Array [
      "rc/components/*/*.svelte",
      "rc/components/*.svelte",
    ],
    "output": Object {
      "dir": "test/public/svelte",
      "entryFileNames": "entry[name]-[hash].js",
      "format": "system",
      "sourcemap": false,
    },
    "plugins": Array [
      Object {
        "options": [Function],
        "pluginName": "rollup-plugin-multi-input",
      },
      Object {
        "name": "replace",
        "renderChunk": [Function],
        "transform": [Function],
      },
      Object {
        "name": "json",
        "transform": [Function],
      },
      Object {
        "generateBundle": [Function],
        "load": [Function],
        "name": "svelte",
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "name": "rollup-plugin-external-globals",
        "transform": [Function],
      },
      Object {
        "buildStart": [Function],
        "generateBundle": [Function],
        "getPackageInfoForId": [Function],
        "load": [Function],
        "name": "node-resolve",
        "resolveId": [Function],
      },
      Object {
        "buildStart": [Function],
        "load": [Function],
        "name": "commonjs",
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "load": [Function],
        "name": "babel",
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "name": "terser",
        "renderChunk": [Function],
      },
    ],
    "treeshake": true,
  },
  Object {
    "cache": true,
    "input": Array [
      "rc/components/*/*.svelte",
      "rc/components/*.svelte",
    ],
    "output": Object {
      "dir": "test/___ELDER___/compiled",
      "exports": "auto",
      "format": "cjs",
    },
    "plugins": Array [
      Object {
        "options": [Function],
        "pluginName": "rollup-plugin-multi-input",
      },
      Object {
        "name": "replace",
        "renderChunk": [Function],
        "transform": [Function],
      },
      Object {
        "name": "json",
        "transform": [Function],
      },
      Object {
        "generateBundle": [Function],
        "load": [Function],
        "name": "svelte",
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "buildStart": [Function],
        "generateBundle": [Function],
        "getPackageInfoForId": [Function],
        "load": [Function],
        "name": "node-resolve",
        "resolveId": [Function],
      },
      Object {
        "buildStart": [Function],
        "load": [Function],
        "name": "commonjs",
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "generateBundle": [Function],
        "name": "css",
        "transform": [Function],
      },
      Object {
        "name": "terser",
        "renderChunk": [Function],
      },
    ],
    "treeshake": true,
  },
  Object {
    "cache": true,
    "input": Array [
      "/Users/lukewatts/oss/elderjs/src/plugins/pluginA/*.svelte",
    ],
    "output": Object {
      "dir": "test/public/svelte",
      "entryFileNames": "entry[name]-[hash].js",
      "format": "system",
      "sourcemap": false,
    },
    "plugins": Array [
      Object {
        "options": [Function],
        "pluginName": "rollup-plugin-multi-input",
      },
      Object {
        "name": "replace",
        "renderChunk": [Function],
        "transform": [Function],
      },
      Object {
        "name": "json",
        "transform": [Function],
      },
      Object {
        "generateBundle": [Function],
        "load": [Function],
        "name": "svelte",
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "name": "rollup-plugin-external-globals",
        "transform": [Function],
      },
      Object {
        "buildStart": [Function],
        "generateBundle": [Function],
        "getPackageInfoForId": [Function],
        "load": [Function],
        "name": "node-resolve",
        "resolveId": [Function],
      },
      Object {
        "buildStart": [Function],
        "load": [Function],
        "name": "commonjs",
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "load": [Function],
        "name": "babel",
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "name": "terser",
        "renderChunk": [Function],
      },
    ],
    "treeshake": true,
  },
  Object {
    "cache": true,
    "input": Array [
      "/Users/lukewatts/oss/elderjs/src/plugins/pluginA/*.svelte",
    ],
    "output": Object {
      "dir": "test/___ELDER___/compiled",
      "exports": "auto",
      "format": "cjs",
    },
    "plugins": Array [
      Object {
        "options": [Function],
        "pluginName": "rollup-plugin-multi-input",
      },
      Object {
        "name": "replace",
        "renderChunk": [Function],
        "transform": [Function],
      },
      Object {
        "name": "json",
        "transform": [Function],
      },
      Object {
        "generateBundle": [Function],
        "load": [Function],
        "name": "svelte",
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "buildStart": [Function],
        "generateBundle": [Function],
        "getPackageInfoForId": [Function],
        "load": [Function],
        "name": "node-resolve",
        "resolveId": [Function],
      },
      Object {
        "buildStart": [Function],
        "load": [Function],
        "name": "commonjs",
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "generateBundle": [Function],
        "name": "css",
        "transform": [Function],
      },
      Object {
        "name": "terser",
        "renderChunk": [Function],
      },
    ],
    "treeshake": true,
  },
  Object {
    "cache": true,
    "input": Array [
      "/Users/lukewatts/oss/elderjs/src/plugins/pluginB/*.svelte",
    ],
    "output": Object {
      "dir": "test/public/svelte",
      "entryFileNames": "entry[name]-[hash].js",
      "format": "system",
      "sourcemap": false,
    },
    "plugins": Array [
      Object {
        "options": [Function],
        "pluginName": "rollup-plugin-multi-input",
      },
      Object {
        "name": "replace",
        "renderChunk": [Function],
        "transform": [Function],
      },
      Object {
        "name": "json",
        "transform": [Function],
      },
      Object {
        "generateBundle": [Function],
        "load": [Function],
        "name": "svelte",
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "name": "rollup-plugin-external-globals",
        "transform": [Function],
      },
      Object {
        "buildStart": [Function],
        "generateBundle": [Function],
        "getPackageInfoForId": [Function],
        "load": [Function],
        "name": "node-resolve",
        "resolveId": [Function],
      },
      Object {
        "buildStart": [Function],
        "load": [Function],
        "name": "commonjs",
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "load": [Function],
        "name": "babel",
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "name": "terser",
        "renderChunk": [Function],
      },
    ],
    "treeshake": true,
  },
  Object {
    "cache": true,
    "input": Array [
      "/Users/lukewatts/oss/elderjs/src/plugins/pluginB/*.svelte",
    ],
    "output": Object {
      "dir": "test/___ELDER___/compiled",
      "exports": "auto",
      "format": "cjs",
    },
    "plugins": Array [
      Object {
        "options": [Function],
        "pluginName": "rollup-plugin-multi-input",
      },
      Object {
        "name": "replace",
        "renderChunk": [Function],
        "transform": [Function],
      },
      Object {
        "name": "json",
        "transform": [Function],
      },
      Object {
        "generateBundle": [Function],
        "load": [Function],
        "name": "svelte",
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "buildStart": [Function],
        "generateBundle": [Function],
        "getPackageInfoForId": [Function],
        "load": [Function],
        "name": "node-resolve",
        "resolveId": [Function],
      },
      Object {
        "buildStart": [Function],
        "load": [Function],
        "name": "commonjs",
        "resolveId": [Function],
        "transform": [Function],
      },
      Object {
        "generateBundle": [Function],
        "name": "css",
        "transform": [Function],
      },
      Object {
        "name": "terser",
        "renderChunk": [Function],
      },
    ],
    "treeshake": true,
  },
]
`;
